multi/stm32l4-multi/flash.c

```diff
@@ -232,6 +244,68 @@ size_t flash_writeData(uint32_t offset, const char *buff, size_t size)
 }
 
 
+static unsigned int _getOptions(void)
+{
+	return *(flash_common.flash + flash_optr);
+}
+
+
+static int _setOptions(unsigned int mask, unsigned int val)
+{
+	int err;
+	unsigned int t;
+
+	if ((err = _flash_wait()) < 0)
+		return err;
+
+	_flash_clearFlags();
+
+	t = *(flash_common.flash + flash_optr) & ~mask;
+	*(flash_common.flash + flash_optr) = t | (mask & val);
+	dataBarier();
+	*(flash_common.flash + flash_cr) |= 1 << 17;
+
+	err = _flash_wait();
+	_flash_clearFlags();
+
+	return err;
+}
+
+
+extern int flash_switchBanks(void)
+{
+	int err;
+	unsigned int bfb2;
+
+	mutexLock(flash_common.lock);
+	bfb2 = !!(_getOptions() & opt_bfb2);
+	_program_unlock();
+	_program_optunlock();
+
+	/* If BFB2==1 and Bank 1 is the active bank, then booting from Bank 2
+	must have failed so leave BFB2 as is and just retry booting */
+	if (!bfb2 || flash_common.activebank != 0)
+		err = _setOptions(opt_bfb2, !bfb2);
+	else
+		err = 0;
+
+	if (!err) {
+		/* Reset by flash option loader */
+		*(flash_common.flash + flash_cr) |= 1 << 27;
+		dataBarier();
+
+		/* Reset failed */
+		err = -1;
+	}
```

Comment:
```suggestion
	if (!bfb2 || flash_common.activebank != 0) {
		err = _setOptions(opt_bfb2, !bfb2);
	}
	else {
		/* Reset by flash option loader */
		*(flash_common.flash + flash_cr) |= 1 << 27;
		dataBarier();

		/* Reset failed */
		err = -1;
	}
```

Reply:
Although your suggestion cleans up code, it also changes the logic - the reset will be performed only when booting from Bank 2 failed. The reset shall actually be performed in all circumstances (except when _setOptions fails).

Reply:
Right